### PR TITLE
Add layout: sidebar and app header

### PR DESCRIPTION
**Pull request by david-mears-2**
### PR description

This PR adds what is common between all pages: an openable/closable sidebar, and an app header. These should work on all kinds of devices.

You can even test it directly on your mobile device using your local network by following the instructions in the README:
```bash
npm run dev -- --host
```

It also introduces github action workflows for testing and linting.

There aren't any component tests for the BreadCrumb component because I'm not yet sure how that will be implemented - treat it as an aspect of styling for now. (I created a separate ticket for that.)

### Pages that exist

The blank 'pages' you'll be able to visit are determined by what's in the `pages/` directory, that is:
* localhost:3000/comparison
* localhost:3000/ -- for pages/index.vue

### CODE REVIEW

**Comment by david-mears-2** [comment]
I reverted out the 'do linting' commit because it completely clutters up the PR diff. I can add it back after reviews, and then the 'lint' workflow will pass, as shown by the green tick on that commit.

**Review by M-Kusumgar — COMMENTED** [review]
Looks good testing on mobile with that QR code is super handy as well! Couple points:
* e2e test was flaky on local, is this something you have experienced? it was the opening and closing sidebar test, kept failing on random browsers every time
* how do i get to comparison page by clicking buttons, im assuming this will happen when you create a couple of scenarios ( so in the future)?

I had a couple questions too!

File: components/AppHeader.vue [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { useRoute } from 'vue-router';
+
+const route = useRoute();
+
+const showBreadcrumbs = computed(() => route.meta.hideBreadcrumbs !== true);
+
+const emit = defineEmits(['toggle-sidebar']);
+const toggleSidebar = () => {
+  emit('toggle-sidebar');
+};
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/AppHeader.vue:47]: Yeah. I think this way, either Vue or Nuxt will perform a check that the event emitted is actually explicitly allowed by the defineEmits, but if it comes directly from the template, it doesn't do that check.

File: components/AppHeader.vue [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        <CNavItem class="py-1">
+          <VerticalRule/>
+        </CNavItem>
+        <CNavItem href="#">
+          <CIcon icon="cilCloudDownload" size="lg"/>
+        </CNavItem>
+        <CNavItem class="py-1">
+          <VerticalRule/>
+        </CNavItem>
+        <CNavItem href="#"> ?</CNavItem>
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/AppHeader.vue:26]: I don't know how it's going missing for you - it should always be present. These are the buttons in the top right of the app header.

File: pages/scenario/new.vue [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<template>
+  <div>
+    <Globe />
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply pages/scenario/new.vue:3]: Yes, or more precisely I'm not including it in the PR, as it's WIP. I should have commented this out.

**Comment by david-mears-2** [comment]
> how do i get to comparison page by clicking buttons, im assuming this will happen when you create a couple of scenarios ( so in the future)?

Yes, there's currently no way to get to a comparison by clicking around. You'll be able to get to it from either a single scenario (which you want to compare), or from your bookmarks.

**Comment by david-mears-2** [comment]
> e2e test was flaky on local, is this something you have experienced? it was the opening and closing sidebar test, kept failing on random browsers every time

Yes it has some flakiness for me locally :( but not as bad as 'every time'. I've just run `npm run test:e2e` 5 times in a row and got no failures/flakys.

File: components/AppHeader.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply components/AppHeader.vue:26]: hmmm thats odd then 
image
can confirm no buttons, are these always present for you? they appear when the screen has smaller width

**Comment by M-Kusumgar** [comment]
> Yes it has some flakiness for me locally :( but not as bad as 'every time'. I've just run `npm run test:e2e` 5 times in a row and got no failures/flakys.

maybe its on my end then, ran it 3 times in a row, got multiple failures on every try

File: components/AppHeader.vue [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    </CContainer>
+  </CHeader>
+</template>
+
+<script setup lang="ts">
+import { CIcon } from '@coreui/icons-vue';
+import throttle from 'lodash.throttle';
+import { useRoute } from 'vue-router';
+
+const route = useRoute();
```
Comment (absternator) [inline components/AppHeader.vue:40]: just a tiny layout thing but may be best to seperate your variables at the top and functions below so there is a clear distinction

File: components/WebsocketConnection.client.vue
Comment (absternator) [inline components/WebsocketConnection.client.vue:1]: is there tests for this file?

File: layouts/default.vue [hunk 5]
```diff
@@ -14,10 +20,30 @@
...
 
 <script setup lang="ts">
+const sidebarVisible = ref(false);
+
+const handleToggleSidebar = () => {
+  sidebarVisible.value = !sidebarVisible.value;
+}
+
+const handleSidebarHidden = () => {
+  // The 'hide' event is emitted by the CoreUI Sidebar component as a callback whenever the sidebar
```
Comment (absternator) [inline layouts/default.vue:30]: dont need comment. pretty clean as you did it.. comments just add more noise

File: components/AppHeader.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/AppHeader.vue:26]: Turns out I just wasn't making my screen wide enough! I'll fix it

**Review by absternator — COMMENTED** [review]
Great work looks amazing!! and slick UI!!! and code lgtm too..  just a few minor comments

- maybe a way to pin the sidebar so full size so it dosent auto collapse? or dependent on size of screen? 
- Following on from above, hide makes more sense to go in collapsed mode and not fully gone? Then on smaller screens bring in hamburger menu
- Breadcrumb styling a bit weird. unsure if spacing/styling or im used to it being below
- Would be great to add formatter asap too.. a .prettierrc file should be suffice... wil save you having to go and format everything later
- didnt notice any tests for socket io? is that another PR?

**Comment by david-mears-2** [comment]
> Would be great to add formatter asap too.. a .prettierrc file should be suffice... wil save you having to go and format everything later

It's using [ESLint Stylistic](https://eslint.style/), via `@antfu/eslint-config` -- see README.md

File: components/AppHeader.vue [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+onMounted(() => {
+  window.addEventListener('scroll', handleScroll);
+});
+
+onBeforeUnmount(() => {
+  window.removeEventListener('scroll', handleScroll);
+});
+
+const VerticalRule = h('div', { class: "vr h-100 mx-2 text-body text-opacity-75" })
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply components/AppHeader.vue:57]: This was to avoid duplication, but actually since I'm no longer reusing it, I can just go back to html

File: components/BreadCrumb.vue [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      {{ item.text }}
+    </CBreadcrumbItem>
+  </CBreadcrumb>
+</template>
+
+<script lang="ts" setup>
+import { useRoute } from 'vue-router'
+const route = useRoute()
+
+const items = route.meta.breadcrumbs || []
```
[the remark this answers is not in the document]
Comment (M-Kusumgar) [reply components/BreadCrumb.vue:21]: yh might be worth parsing the route in a breadcrumbs component or something like that, probably for another pr though!

**Comment by david-mears-2** [comment]
I worked out an alternative way the sidebar can operate based on @absternator 's suggestion: https://github.com/j-idea/daedalus-web-app/pull/4

**Comment by david-mears-2** [comment]
> Breadcrumb styling a bit weird. unsure if spacing/styling or im used to it being below

What do you mean by 'below'?

**Comment by david-mears-2** [comment]
> didnt notice any tests for socket io? is that another PR?

I will probably only add tests for websockets once we start using them for something. The websockets-related code in this PR is all either setup (copied from docs), or a debugging tool that just proves to the developer that websockets is working.

**Comment by EmmaLRussell** [comment]
> I worked out an alternative way the sidebar can operate based on @absternator 's suggestion: #4

Yeah, new way definitely seems preferable. Shouldn't need the hamburger icon on larger screens though, since collapsed view always visible?

**Comment by david-mears-2** [comment]
> Shouldn't need the hamburger icon on larger screens though, since collapsed view always visible?

In the new way, I made it so the hamburger toggles the width of the sidebar - it switches between:
* Fixed open sidebar
and
* Sidebar that opens on hover

So it's about whether we think people will want to be able to fix it open.

**Comment by EmmaLRussell** [comment]
> So it's about whether we think people will want to be able to fix it open.

Oh I see - well, they might do, but I'm not sure the hamburger icon is the way to offer that functionality, since it's customary use is to open. Something like a lock icon maybe, but probably something on the sidebar (like your original Hide control) would have been my first thought.

File: components/AppHeader.vue [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+onMounted(() => {
+  window.addEventListener('scroll', handleScroll)
+})
+
+onBeforeUnmount(() => {
+  window.removeEventListener('scroll', handleScroll)
+})
+</script>
+
+<template>
```
Comment (EmmaLRussell) [inline components/AppHeader.vue:31]: I've always preferred putting the template above the script, as it feels more natural to see your static markup before the dynamic operations which work upon it. But again, this is quite subjectve!

**Comment by david-mears-2** [comment]
I pushed a new commit to https://github.com/j-idea/daedalus-web-app/pull/4 that removes the toggle for large screens.

File: pages/comparison.vue [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+<script lang="ts" setup>
+// A non-dynamic placeholder example of what the breadcrumb text would be.
+// The text should actually depend on the route (which determines which axis is the axis of comparison).
+definePageMeta({
+  breadcrumbs: [
+    { text: 'Base scenario', href: '#' },
+    { text: 'Explore more scenarios', href: '#' },
+    { text: `By policy response` }
+  ]
+})
```
Comment (EmmaLRussell) [inline pages/comparison.vue:16]: Yeah, it feels like the router ought to be giving us breadcrumb information.

File: pages/index.vue [hunk 10]
```diff
@@ -3,3 +3,12 @@
     <h1 style="height: 5000px;">Home page</h1>
   </div>
 </template>
+
+<script setup lang="ts">
+definePageMeta({
+  hideBreadcrumbs: true
```
Comment (EmmaLRussell) [inline pages/index.vue:9]: ..again, it feels like the breadcrumb component itself might make decision about hiding breadbrumb if only one level of route etc, rather than page defining it in the meta.

**Review by EmmaLRussell — COMMENTED** [review]
Looks great - really slick. 

I'm not completely convinced by having the logos appear in the sidebar, that Imperial "I" is quite distracting... but I guess they have to go somewhere. We put them on the About page in arbomap - will you need one of those too? - so maybe could go there if acceptable?

The ? is going to be the Help link? It would look more consistent to make that a logo too I think, and give it a tooltip. 

I think the breadcrumbs would look a little cleaner without underline for links, and without the grey horizontal line between header and breadcrumb on smaller screen. But that's very subjective...

Unit tests for components going in a separate ticket?

File: components/AppHeader.vue (hunk 8 above)
Comment (david-mears-2) [reply components/AppHeader.vue:31]: I agree actually, so I'm now overriding the lint/format rule that caused this.

File: components/AppHeader.vue (hunk 4 above)
Comment (david-mears-2) [reply components/AppHeader.vue]: Hopefully this will be sorted out by https://github.com/j-idea/daedalus-web-app/pull/2#issuecomment-2225919107

**Comment by david-mears-2** [comment]
> I'm not completely convinced by having the logos appear in the sidebar, that Imperial "I" is quite distracting... but I guess they have to go somewhere. We put them on the About page in arbomap - will you need one of those too? - so maybe could go there if acceptable?

I'm not sure if we will have an 'About' page in so many words. We'll have a page that explains methodology and links to the paper, which is similar, but might not get called 'About' and I'm not sure the logo/acknowledgements should go on that page.

I've been led to understand that branding it with the name Jameel is important. For now, I will remove the Imperial I so the logo only shows when the sidebar is open (I wasn't convinced about my solution for the distractingness of the I - I was reducing its opacity).

**Comment by david-mears-2** [comment]
> The ? is going to be the Help link? It would look more consistent to make that a logo too I think, and give it a tooltip.

Agreed. I didn't because there wasn't such an icon in the library I was using. Tooltip (you mean just the word 'Help' or something?) comes in https://github.com/j-idea/daedalus-web-app/pull/3. I'm going to add an svg from font awesome to assets.

**Comment by david-mears-2** [comment]
I think I've addressed all comments. I merged into this branch the new way for sidebars to work.

I intend to work out how we 'implement breadcrumbs properly' in [this ticket](https://mrc-ide.myjetbrains.com/youtrack/issue/JIDEA-32/Implement-bread-crumbs-properly).

After final reviews, I'll run `lint:fix` which will do all the formatting and linting, but I won't do that for now because it will make things unreadable.

**Review by EmmaLRussell — COMMENTED** [review]
Looks good! Think I preferred the breadcrumb left/center rather than on the right.  

When you enable the formatter/linter, can you make it consistent with our other web apps?
i.e. double quotes, semicolon line endings. We're not completely consistent on indents - I prefer 4 spaces, but we also have 2 spaces...

File: tests/components/SideBar.spec.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+import { describe, expect, it } from 'vitest'
+import { mountSuspended } from '@nuxt/test-utils/runtime'
+import SideBar from '@/components/SideBar.vue'
+
+const stubs = {
+  CIcon: true,
+}
+
+// Reviewers - I wonder if these tests are too brittle
+// (tied closely to the component's implementation / would change too often)
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply tests/components/SideBar.spec.ts:12]: Hmm, well my thoughts are that I already have e2e tests on the overall behavior of the sidebar, making my below tests of the implementation somewhat redundant -- except in that unit tests give quicker feedback about regressions (not only that they are quicker to run but also quicker to pinpoint what broke).

**Review by absternator — COMMENTED** [review]
Nice work LGTM.. i wont approve as you and emma still ironing things out :smile:

File: layouts/default.vue [hunk 12]
```diff
@@ -20,30 +20,74 @@
...
-};
+function handleSidebarHidden() {
+  if (hideHasBeenEmitted.value) {
+    // If this is not the 'hide' emitted on page load, we obey it.
+    sidebarVisible.value = false
+  }
+  else {
+    // If this is the 'hide' emitted on page load, we un-do it.
+    // This is because the CoreUI Sidebar component emits a 'hide' event on page load, which we
+    // don't want to obey for larger screen sizes.
```
[the remark this answers is not in the document]
Comment (david-mears-2) [reply layouts/default.vue]: Good point. I've now refactored this, using [defineModel()](https://vuejs.org/guide/components/v-model)

File: tests/components/SideBar.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply tests/components/SideBar.spec.ts]: I think unit tests are good for fine detailed test work - you might test edge cases that just involve that component in unit. e2e tests should test the main use cases work end to end, and any behaviours that require multiple components to coordinate in complex ways.

**Comment by EmmaLRussell** [comment]
Code looking good to me now - did you want to fix up the lint build before merging?
